File: backend/app/schemas/selfie.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


ModerationMode = Literal["auto_approve", "manual_approve"]
OverlayMode = Literal["logo", "qr", "off"]


class SelfieStateRead(BaseModel):
    slideshow_enabled: bool
    slideshow_interval_seconds: int
    slideshow_max_visible_photos: int
    slideshow_min_uploads_to_start: int
    slideshow_shuffle: bool
    overlay_mode: OverlayMode
    vintage_look_enabled: bool
    moderation_mode: ModerationMode
    slideshow_updated_at: datetime | None


class SelfieStateUpdate(BaseModel):
    slideshow_enabled: bool
    slideshow_interval_seconds: int = Field(ge=2, le=20)
    slideshow_max_visible_photos: int = Field(ge=1, le=10)
    slideshow_min_uploads_to_start: int = Field(ge=1, le=50)
    slideshow_shuffle: bool
    overlay_mode: OverlayMode
    vintage_look_enabled: bool
    moderation_mode: ModerationMode

File: frontend/src/stores/selfie.ts
import { ref } from 'vue'
import { defineStore } from 'pinia'

import type { SelfieState } from '../services/api'
import { fetchSelfieState, updateSelfieState } from '../services/api'

export const useSelfieStore = defineStore('selfie', () => {
  const slideshowEnabled = ref(true)
  const slideshowIntervalSeconds = ref(6)
  const slideshowMaxVisiblePhotos = ref(4)
  const slideshowMinUploadsToStart = ref(3)
  const slideshowShuffle = ref(true)
  const overlayMode = ref<SelfieState['overlay_mode']>('logo')
  const vintageLookEnabled = ref(false)
  const moderationMode = ref<SelfieState['moderation_mode']>('auto_approve')
  const slideshowUpdatedAt = ref<string | null>(null)

  async function refresh() {
    applyState(await fetchSelfieState())
  }

  async function save(nextState: Omit<SelfieState, 'slideshow_updated_at'>) {
    const token = localStorage.getItem('raveberg-admin-token')
    if (!token) {
      throw new Error('Admin session required')
    }

    applyState(await updateSelfieState(nextState, token))
  }

  function applyState(state: SelfieState) {
    slideshowEnabled.value = state.slideshow_enabled
    slideshowIntervalSeconds.value = state.slideshow_interval_seconds
    slideshowMaxVisiblePhotos.value = state.slideshow_max_visible_photos
    slideshowMinUploadsToStart.value = state.slideshow_min_uploads_to_start
    slideshowShuffle.value = state.slideshow_shuffle
    overlayMode.value = state.overlay_mode
    vintageLookEnabled.value = state.vintage_look_enabled
    moderationMode.value = state.moderation_mode
    slideshowUpdatedAt.value = state.slideshow_updated_at
  }

  return {
    slideshowEnabled,
    slideshowIntervalSeconds,
    slideshowMaxVisiblePhotos,
    slideshowMinUploadsToStart,
    slideshowShuffle,
    overlayMode,
    vintageLookEnabled,
    moderationMode,
    slideshowUpdatedAt,
    refresh,
    save,
    applyState,
  }
})
